File: CloudServer/src/DeviceManager/Sessions.go
package DeviceManager

import ()

type SessionManager struct {
	activeSessions []uint16
}

//instantiate a global session manager
var GSM SessionManager = SessionManager{}

func (sm *SessionManager) IsPresent(x uint16) int {
	for i := range sm.activeSessions {
		if sm.activeSessions[i] == x {
			return i
		}
	}
	return -1
}

func (sm *SessionManager) Add(x uint16) bool {
	if sm.IsPresent(x) == -1 {
		//fmt.Println("Adding new ")
		sm.activeSessions = append(sm.activeSessions, []uint16{x}...)
		return true
	} else {
		return false
	}
}

func (sm *SessionManager) Delete(x uint16) bool {
	i := sm.IsPresent(x)
	if i == -1 {
		return false
	} else {
		sm.activeSessions = append(sm.activeSessions[:i], sm.activeSessions[i+1:]...)
		return true
	}
}

func (sm *SessionManager) Length() int {
	return len(sm.activeSessions)
}

File: MicroFrameworkPK_v4_3/DSM/ComManager/Sessions.cpp
#include "Sessions.h"


SessionManager GSM;

bool SessionManager::Add(UINT16 no){
	if(IsPresent(no)){
		return false;
	}
	if(numberOfSessions>=MAX_SESSIONS) return false;
	sess[numberOfSessions]=no;
	numberOfSessions++;
}
bool SessionManager::IsPresent(UINT16 no){
	for (int i=0; i<MAX_SESSIONS; i++){
		if(sess[i]==no){
			return true;
		}
	}
	return false;
}
bool SessionManager::Delete(UINT16 no){
	for (int i=0; i<MAX_SESSIONS; i++){
		if(sess[i]==no){
			if(i<numberOfSessions){
				sess[i]=sess[numberOfSessions];
			}else {
				sess[i]=0;
			}
			numberOfSessions--;
			return true;
		}
	}
	return false;

}

UINT8 SessionManager::Length(){
	return numberOfSessions;
}

void SessionManager::Init(){
	for (int i=0; i<MAX_SESSIONS; i++){
		sess[i]=0;
	}
	numberOfSessions=0;
}
